CBDatabase/DB/CoreDataDiskStorage.swift

```diff
@@ -0,0 +1,258 @@
+                attributes: nil
+            )
+        }
+
+        let storeURL = docURL.appendingPathComponent("\(options.dbStorageFilename).sqlite")
+        var previousManagerObjectModel: NSManagedObjectModel?
+        let currentMigrationFilename = options.currentMigrationVersionFilename
+        let currentVersion = currentMigrationVersion(filename: currentMigrationFilename)
+        let managedObjectModels: [(String, NSManagedObjectModel)] = try options.versions.compactMap { version in
+            let resource = "\(options.dbSchemaName).momd/\(version)"
+
+            guard
+                let versionURL = options.dataModelBundle.url(forResource: resource, withExtension: "mom"),
+                let managedObjectModel = NSManagedObjectModel(contentsOf: versionURL)
+            else {
+                throw DatabaseError.unableToCreateManagedObjectModel
+            }
+
+            return (version, managedObjectModel)
+        }
+
+        let startIndex = managedObjectModels.firstIndex(where: { $0.0 == currentVersion }) ?? 0
+        try managedObjectModels[startIndex...].forEach { versionName, mom1 in
+            let metadata = try? NSPersistentStoreCoordinator.metadataForPersistentStore(
+                ofType: NSSQLiteStoreType,
+                at: storeURL
+            )
+
+            if
+                let aMetadata = metadata,
+                !mom1.isConfiguration(withName: nil, compatibleWithStoreMetadata: aMetadata),
+                let mom0 = previousManagerObjectModel {
+                let bundle = options.dataModelBundle
+                try migrate(from: mom0, to: mom1, storeURL: storeURL, bundle: bundle)
+            } else {
+                try setup(managedObjectModel: mom1, storeURL: storeURL)
+            }
+
+            try setCurrentMigrationVersion(versionName, filename: currentMigrationFilename)
+            previousManagerObjectModel = mom1
+        }
+
+        guard let recentManagedObjectModels = managedObjectModels.last?.1 else {
+            throw DatabaseError.unableToSetupDatabase
+        }
+
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        let psc = try setup(managedObjectModel: recentManagedObjectModels, storeURL: storeURL)
+
+        managedObjectContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        managedObjectContext.persistentStoreCoordinator = psc
+
+        return managedObjectContext
+    }
+
+    private static func currentMigrationVersion(filename: String) -> String? {
+        guard
+            let fileURL = docURL?.appendingPathComponent(filename),
+            FileManager.default.fileExists(atPath: fileURL.path),
+            let version = try? String(contentsOf: fileURL, encoding: .utf8)
+        else { return nil }
+
+        return version
+    }
+
+    private static func setCurrentMigrationVersion(_ versionName: String?, filename: String) throws {
```

Comment:
not required but do you think it's worth making a `MigrationVersion` type that has a filename and version properties? unsure if this will clean it up or make things more complicated but seems like they're paired up in a number of places

Reply:
The reason I don't have them together in one model is because the filename is static and depends on the initial DB store name. i.e see `DiskDatabaseOptions.currentMigrationVersionFilename`